### PR TITLE
create a pubsub mechanism through redis for indexer to dong #876

### PR DESCRIPTION
**Pull request by QTR33**
# Ticket
#876

# General
- [ ] No unnecessary files, debug logs, or commented code included.
- [ ] PR has at least 1 peer approval before senior review.
- [ ] I fully understand the code I submitted.
- [ ] No hardcoded configuration or secrets.
- [ ] **Environment Variables** changes? (Leave unchecked if N/A)

# Design & Solution Clarity
- [ ] A design diagram (sequence / flow) is attached or linked.
- [ ] The implementation matches the design.
- [ ] All major flows and decision points are represented in the diagram.
- [ ] Error handling and edge cases are clearly defined in the design.

Design link (diagram / doc):
`<paste link here>`

# Testing
- [ ] Unit tests cover main logic and important edge cases.

> Note: Complex logic must not rely on manual testing only.

# UI evidence (For frontend only)
`<paste screenshots here>`

# Peer Review Checklist
- [ ] I understand the overall solution and flow.
- [ ] The code follows the provided design diagram.
- [ ] No obvious logic or security issues found.
- [ ] Tests are sufficient for the scope of this change.

### CODE REVIEW

File: dong-service/services/indexer_sync_service.go [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+		query := fmt.Sprintf("UPDATE %s.red_envelope SET status = $1, transaction_hash = $2, updated_at = NOW() WHERE id = $3", s.dongSchema)
+		if _, err = tx.ExecContext(ctx, query, u.Status, u.TxHash, u.ID); err != nil {
+			logger.Error().Err(err).Str("re_id", u.ID).Msg("Failed to update red envelope status from sync")
+			return err
+		}
+	}
+
+	// 4. Process Red Envelope Claim Updates
+	for _, u := range msg.RedEnvelopeClaimUpdates {
+		query := fmt.Sprintf("UPDATE %s.red_envelope_claim SET status = $1, transaction_hash = $2 WHERE id = $3", s.dongSchema)
```
Comment (vietdangquoc) [inline dong-service/services/indexer_sync_service.go]: updated_at = NOW()

File: dong-service/services/indexer_sync_service.go [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+	defer func() {
+		if err != nil {
+			_ = tx.Rollback()
+		}
+	}()
+
+	// 1. Process Offer Updates
+	for _, u := range msg.OfferUpdates {
+		if err = s.offerRepo.UpdateOfferStatus(ctx, u.OfferID, u.Status, tx, &u.TxHash); err != nil {
```
Comment (vietdangquoc) [inline dong-service/services/indexer_sync_service.go:58]: kiểm tra lại xem có missing socket event không

Comment (vietdangquoc) [reply dong-service/services/indexer_sync_service.go:58]: hoặc phải thêm socket event bên services

File: indexer/internal/storage/pubsub.go [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+type RedEnvelopeUpdate struct {
+	ID     string `json:"id"`
+	Status string `json:"status"`
+	TxHash string `json:"tx_hash"`
+}
+
+// PublishDongUpdates sends a batch of updates to the Dong Service via Redis Pub/Sub
+func PublishDongUpdates(ctx context.Context, msg *DongBatchMessage) error {
+	if RedisClient == nil {
```
Comment (vietdangquoc) [inline indexer/internal/storage/pubsub.go:49]: add log

File: indexer/internal/storage/postgres.go [hunk 4]
```diff
@@ -2857,47 +2817,34 @@ func (p *PostgresConnector) updateOrderStatus(
 
 		validOrderIDs = append(validOrderIDs, o.OrderID)
 		validTxHashes = append(validTxHashes, t.Hash)
+		updates = append(updates, OrderUpdate{
+			OfferID: o.OfferID,
+			OrderID: o.OrderID,
+			Status:  "OPEN", // Wait, status in repo is 'OPEN' but for order maybe it means 'COMPLETED'?
```
Comment (vietdangquoc) [inline indexer/internal/storage/postgres.go]: ??? comment gì zậy
kiểm tra lại status

File: dong-service/repository/intermediary_wallet_repository.go [hunk 5]
```diff
@@ -325,11 +325,21 @@ func (r *IntermediaryWalletRepository) UpdateWalletStatus(ctx context.Context, w
 }
 
 func (r *IntermediaryWalletRepository) ReleaseWallet(ctx context.Context, walletAddress string) error {
```
Comment (vietdangquoc) [inline dong-service/repository/intermediary_wallet_repository.go:327]: sao k gộp luôn ReleaseWalletWithTx

File: indexer/internal/storage/postgres.go (hunk 4 above)
Comment (QTR33) [reply indexer/internal/storage/postgres.go]: done

File: indexer/internal/storage/pubsub.go (hunk 3 above)
Comment (QTR33) [reply indexer/internal/storage/pubsub.go:49]: done

File: dong-service/services/indexer_sync_service.go (hunk 1 above)
Comment (QTR33) [reply dong-service/services/indexer_sync_service.go]: done

File: dong-service/repository/intermediary_wallet_repository.go (hunk 5 above)
Comment (QTR33) [reply dong-service/repository/intermediary_wallet_repository.go:327]: cái này chia làm 2 cách dùng có transaction và không dùng transaction nên e để như này ạ

File: dong-service/services/indexer_sync_service.go (hunk 2 above)
Comment (QTR33) [reply dong-service/services/indexer_sync_service.go:58]: done
